File: Editorial/79_Word_Search/C/1_Backtracking.c
bool exist(char** board, int boardSize, int* boardColSize, char* word) {
    int ROWS = boardSize;
    int COLS = *boardColSize;
    bool backtrack(int row, int col, char* suffix) {
        /* Step 1). check the bottom case. */
        if (strlen(suffix) == 0) return true;
        /* Step 2). Check the boundaries. */
        if (row < 0 || row == ROWS || col < 0 || col == COLS ||
            board[row][col] != suffix[0])
            return false;
        /* Step 3). explore the neighbors in DFS */
        bool ret = false;
        // mark the path before the next exploration
        board[row][col] = '#';
        int directions[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for (int d = 0; d < 4; ++d) {
            ret = backtrack(row + directions[d][0], col + directions[d][1],
                            suffix + 1);
            if (ret) break;
        }
        /* Step 4). clean up and return the result. */
        board[row][col] = suffix[0];
        return ret;
    }

    for (int row = 0; row < ROWS; ++row) {
        for (int col = 0; col < COLS; ++col) {
            if (backtrack(row, col, word)) return true;
        }
    }
    return false;
};

File: Editorial/79_Word_Search/JavaScript/1_Backtracking.js
var exist = function (board, word) {
    const ROWS = board.length;
    const COLS = board[0].length;
    const backtrack = function (row, col, suffix) {
        /* Step 1). check the bottom case. */
        if (suffix.length == 0) return true;
        /* Step 2). Check the boundaries. */
        if (
            row < 0 ||
            row == ROWS ||
            col < 0 ||
            col == COLS ||
            board[row][col] != suffix.charAt(0)
        )
            return false;
        /* Step 3). explore the neighbors in DFS */
        let ret = false;
        // mark the path before the next exploration
        board[row][col] = "#";
        const directions = [
            [0, 1],
            [1, 0],
            [0, -1],
            [-1, 0],
        ];
        for (let [rowOffset, colOffset] of directions) {
            ret = backtrack(row + rowOffset, col + colOffset, suffix.slice(1));
            if (ret) break;
        }
        /* Step 4). clean up and return the result. */
        board[row][col] = suffix.charAt(0);
        return ret;
    };

    for (let row = 0; row < ROWS; ++row) {
        for (let col = 0; col < COLS; ++col) {
            if (backtrack(row, col, word)) return true;
        }
    }
    return false;
};
